File: backend/src/main/java/com/horain/agent/ExportEvalCandidatesService.java
package com.horain.agent;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.horain.model.AgentFeedback;
import com.horain.model.AgentTurn;
import com.horain.repository.AgentFeedbackRepository;
import com.horain.repository.AgentTurnRepository;
import org.springframework.stereotype.Service;

import java.util.*;
import java.util.stream.Collectors;

/**
 * Builds eval candidates (turns with thumbs-down or tool/empty errors) for export.
 * Used by the admin HTTP endpoint and optionally by the CLI export runner.
 */
@Service
public class ExportEvalCandidatesService {

    private final AgentTurnRepository turnRepository;
    private final AgentFeedbackRepository feedbackRepository;
    private final ObjectMapper objectMapper;

    public ExportEvalCandidatesService(AgentTurnRepository turnRepository,
                                       AgentFeedbackRepository feedbackRepository,
                                       ObjectMapper objectMapper) {
        this.turnRepository = turnRepository;
        this.feedbackRepository = feedbackRepository;
        this.objectMapper = objectMapper;
    }

    /**
     * Returns one row (map) per eval candidate, sorted by created_at descending.
     */
    public List<Map<String, Object>> getCandidates() {
        Set<UUID> turnIds = new LinkedHashSet<>();
        List<AgentFeedback> downFeedbacks = feedbackRepository.findByRating("down");
        for (AgentFeedback f : downFeedbacks) {
            turnIds.add(f.getTurnId());
        }
        List<AgentTurn> errorTurns = turnRepository.findByStatusIn(List.of("tool_error", "empty_result", "max_iterations"));
        for (AgentTurn t : errorTurns) {
            turnIds.add(t.getId());
        }

        Map<UUID, AgentFeedback> feedbackByTurn = downFeedbacks.stream()
                .collect(Collectors.toMap(AgentFeedback::getTurnId, f -> f, (a, b) -> a));

        List<AgentTurn> turns = new ArrayList<>();
        for (UUID id : turnIds) {
            turnRepository.findById(id).ifPresent(turns::add);
        }
        turns.sort(Comparator.comparing(AgentTurn::getCreatedAt).reversed());

        List<Map<String, Object>> rows = new ArrayList<>();
        for (AgentTurn turn : turns) {
            Map<String, Object> row = new LinkedHashMap<>();
            row.put("source_turn_id", turn.getId().toString());
            row.put("conversation_id", turn.getConversationId().toString());
            row.put("user_message", turn.getUserMessage());
            row.put("assistant_message", turn.getAssistantMessage());
            row.put("system_prompt_version", turn.getSystemPromptVersion());
            row.put("model", turn.getModel());
            row.put("status", turn.getStatus());
            row.put("created_at", turn.getCreatedAt() != null ? turn.getCreatedAt().toString() : null);
            if (turn.getToolCallsJson() != null && !turn.getToolCallsJson().isBlank()) {
                try {
                    @SuppressWarnings("unchecked")
                    List<Map<String, String>> toolCalls = objectMapper.readValue(turn.getToolCallsJson(), List.class);
                    row.put("tool_calls", toolCalls);
                } catch (Exception e) {
                    row.put("tool_calls_raw", turn.getToolCallsJson());
                }
            }
            AgentFeedback fb = feedbackByTurn.get(turn.getId());
            if (fb != null) {
                row.put("feedback", fb.getRating());
                row.put("feedback_reason", fb.getReasonCode());
                row.put("feedback_comment", fb.getComment());
            } else {
                row.put("feedback", null);
            }
            row.put("expected_behavior", "");
            row.put("eval_family", "");
            row.put("assertion_strategy", "deterministic");
            rows.add(row);
        }
        return rows;
    }

    public ObjectMapper getObjectMapper() {
        return objectMapper;
    }
}

File: backend/src/main/kotlin/com/horain/agent/ExportEvalCandidatesService.kt
package com.horain.agent

import com.fasterxml.jackson.databind.ObjectMapper
import com.horain.repository.AgentFeedbackRepository
import com.horain.repository.AgentTurnRepository
import org.springframework.stereotype.Service
import java.util.LinkedHashMap
import java.util.UUID

/**
 * Builds eval candidates (turns with thumbs-down or tool/empty errors) for export.
 * Used by the admin HTTP endpoint and optionally by the CLI export runner.
 */
@Service
class ExportEvalCandidatesService(
    private val turnRepository: AgentTurnRepository,
    private val feedbackRepository: AgentFeedbackRepository,
    val objectMapper: ObjectMapper
) {

    /**
     * Returns one row (map) per eval candidate, sorted by created_at descending.
     */
    fun getCandidates(): List<Map<String, Any?>> {
        val turnIds = linkedSetOf<UUID>()
        val downFeedbacks = feedbackRepository.findByRating("down")
        for (f in downFeedbacks) {
            turnIds.add(f.turnId!!)
        }
        val errorTurns = turnRepository.findByStatusIn(
            listOf("tool_error", "empty_result", "max_iterations")
        )
        for (t in errorTurns) {
            turnIds.add(t.id!!)
        }
        val feedbackByTurn = downFeedbacks.associateBy { it.turnId!! }
        val turns = mutableListOf<com.horain.model.AgentTurn>()
        for (id in turnIds) {
            turnRepository.findById(id).ifPresent { turns.add(it) }
        }
        turns.sortByDescending { it.createdAt }
        val rows = mutableListOf<Map<String, Any?>>()
        for (turn in turns) {
            val row = LinkedHashMap<String, Any?>()
            row["source_turn_id"] = turn.id.toString()
            row["conversation_id"] = turn.conversationId.toString()
            row["user_message"] = turn.userMessage
            row["assistant_message"] = turn.assistantMessage
            row["system_prompt_version"] = turn.systemPromptVersion
            row["model"] = turn.model
            row["status"] = turn.status
            row["created_at"] = turn.createdAt?.toString()
            if (!turn.toolCallsJson.isNullOrBlank()) {
                try {
                    @Suppress("UNCHECKED_CAST")
                    val toolCalls = objectMapper.readValue(turn.toolCallsJson, List::class.java) as List<Map<String, String>>
                    row["tool_calls"] = toolCalls
                } catch (_: Exception) {
                    row["tool_calls_raw"] = turn.toolCallsJson
                }
            }
            val fb = feedbackByTurn[turn.id]
            if (fb != null) {
                row["feedback"] = fb.rating
                row["feedback_reason"] = fb.reasonCode
                row["feedback_comment"] = fb.comment
            } else {
                row["feedback"] = null
            }
            row["expected_behavior"] = ""
            row["eval_family"] = ""
            row["assertion_strategy"] = "deterministic"
            rows.add(row)
        }
        return rows
    }
}
